File: src/dataset.py
import csv
from dataclasses import dataclass
from typing import List


@dataclass
class Event:
    time: int
    process_id: str
    action: str
    resource_id: str
    duration: int


def load_events(path: str) -> List[Event]:
    events = []
    with open(path, newline="") as f:
        reader = csv.DictReader(f)
        for row in reader:
            events.append(
                Event(
                    time=int(row["time"]),
                    process_id=row["process_id"],
                    action=row["action"],
                    resource_id=row["resource_id"],
                    duration=int(row["duration"]),
                )
            )
    events.sort(key=lambda e: e.time)
    return events

File: src/dataset.c
#include "dataset.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static int cmp_events(const void* a, const void* b) {
    return ((const Event*)a)->time - ((const Event*)b)->time;
}

Event* load_events(const char* path, int* count) {
    FILE* f = fopen(path, "r");
    if (!f) { *count = 0; return NULL; }

    int capacity = 512;
    Event* events = (Event*)malloc((size_t)capacity * sizeof(Event));
    *count = 0;

    char line[256];
    fgets(line, (int)sizeof(line), f);

    while (fgets(line, (int)sizeof(line), f)) {
        if (*count >= capacity) {
            capacity *= 2;
            events = (Event*)realloc(events, (size_t)capacity * sizeof(Event));
        }
        Event* e = &events[*count];
        if (sscanf(line, "%d,%15[^,],%15[^,],%15[^,],%d",
                   &e->time, e->process_id, e->action, e->resource_id, &e->duration) == 5)
            (*count)++;
    }
    fclose(f);
    // Dataset is already sorted by time, no need to qsort and risk unstable sorting!
    return events;
}

void free_events(Event* events) {
    free(events);
}
